libs/plugin/src/generators/convert-to-sfc/generator.ts

```diff
@@ -0,0 +1,171 @@
+import { dirname } from 'node:path';
+import { exit } from 'node:process';
+import { Node, SyntaxKind } from 'ts-morph';
+import { ContentsStore } from '../shared-utils/contents-store';
+import { ConvertToSFCGeneratorSchema } from './schema';
+
+function trackContents(
+	tree: Tree,
+	contentsStore: ContentsStore,
+	fullPath: string,
+) {
+	if (fullPath.endsWith('.ts')) {
+		const fileContent =
+			tree.read(fullPath, 'utf8') || readFileSync(fullPath, 'utf8');
+		if (!fileContent.includes('@Component')) return;
+		if (fileContent.includes('templateUrl')) {
+			contentsStore.track(fullPath, fileContent);
+		}
+	}
+}
+
+export async function convertToSFCGenerator(
+	tree: Tree,
+	options: ConvertToSFCGeneratorSchema,
+) {
+	const contentsStore = new ContentsStore();
+	const packageJson = readJson(tree, 'package.json');
+	const angularCorePackage =
+		packageJson['dependencies']['@angular/core'] ||
+		packageJson['devDependencies']['@angular/core'];
+
+	if (!angularCorePackage) {
+		logger.error(`[ngxtension] No @angular/core detected`);
+		return exit(1);
+	}
+
+	const { path, project, moveStyles, maxInlineTemplateLines } = options;
+
+	if (path && project) {
+		logger.error(
+			`[ngxtension] Cannot pass both "path" and "project" to convertToSFCGenerator`,
+		);
+		return exit(1);
+	}
+
+	if (path) {
+		if (!tree.exists(path)) {
+			logger.error(`[ngxtension] "${path}" does not exist`);
+			return exit(1);
+		}
+
+		trackContents(tree, contentsStore, path);
+	} else if (project) {
+		try {
+			const projectConfiguration = readProjectConfiguration(tree, project);
+
+			if (!projectConfiguration) {
+				throw `"${project}" project not found`;
+			}
+
+			visitNotIgnoredFiles(tree, projectConfiguration.root, (path) => {
+				trackContents(tree, contentsStore, path);
+			});
+		} catch (err) {
+			logger.error(`[ngxtension] ${err}`);
+			return;
+		}
+	} else {
+		const projects = getProjects(tree);
+		for (const project of projects.values()) {
+			visitNotIgnoredFiles(tree, project.root, (path) => {
+				trackContents(tree, contentsStore, path);
+			});
+		}
+	}
+
+	for (const { path: sourcePath } of contentsStore.collection) {
+		if (!sourcePath.endsWith('.ts')) continue;
+
+		const sourceFile = contentsStore.project.getSourceFile(sourcePath)!;
+
+		const classes = sourceFile.getClasses();
+
+		for (const targetClass of classes) {
+			const applicableDecorator = targetClass.getDecorator((decoratorDecl) => {
+				return ['Component'].includes(decoratorDecl.getName());
+			});
+			if (!applicableDecorator) continue;
+
+			const decoratorArg = applicableDecorator.getArguments()[0];
+			if (Node.isObjectLiteralExpression(decoratorArg)) {
+				decoratorArg
+					.getChildrenOfKind(SyntaxKind.PropertyAssignment)
+					.forEach((property) => {
+						const decoratorPropertyName = property.getName();
+						if (decoratorPropertyName === 'templateUrl') {
+							const dir = dirname(sourcePath);
+							const templatePath = joinPathFragments(
```

Comment:
Is it possible to create a reusable function for preparing "templatePath", and then reading its content? I also noticed a similar code in the "self-closing tags" schematic.

Reply:
Good idea. Will do as a refactoring step.